Linux Compatability

tiferrei opened the pull request:
Open discussion for solutions to #54.

tiferrei committed:
AnyObject -> Any

tiferrei committed:
Forgot one

pietbrauer commented:
I think it would also be good to add another travis target that builds on Linux.

tiferrei commented:
@pietbrauer definitely yes, I am also working on making the Unit Tests Linux compatible so that we can just run `swift test` in the CI.

tiferrei commented:
@pietbrauer we have a slight issue with using `URLSession.shared()` in the tests, [it hasn't been implemented yet.](https://github.com/apple/swift-corelibs-foundation/blob/93d0018a0891f876065651c3ee8c63f256a38de4/Foundation/NSURLSession/NSURLSession.swift#L200)
For some reason, it also complains about it on the Swift CLI on macOS.

tiferrei committed:
Make FS more SwiftPM like.

pietbrauer commented:
Hm, not sure if it is worth it then. Do you have a need for it to work on Linux?

tiferrei commented:
@pietbrauer that’s the thing, it works on Linux! It compiles, the _tests_ don’t compile. I’m even working on a server-side Swift project that uses RequestKit and everything works like a charm on Linux and Docker. It’s just the tests for RequestKit that are a bit messed up. The weird part is that they also fail on macOS? On Xcode everything rolls smoothly but not with `swift test`.

tiferrei commented:
@pietbrauer any idea what the CI is complaining about? Seems to be platform specific. If I add the parentheses then the other platforms start complaining

tiferrei committed:
Workaround Swift 3 API inconsistency

tiferrei commented:
@pietbrauer do we have any plans on upgrading to Swift 4? I think a lot of these things have been improved on Swift 4, mainly API consistency between Darwin and Linux.

pietbrauer committed:
Build on Linux

pietbrauer committed:
Make tests compile using Compile Swift Module 'RequestKit' (3 sources)
Compile Swift Module 'RequestKitTests' (7 sources)
/Users/pietbrauer/Projects/RequestKit/Tests/RequestKitTests/ConfigurationTests.swift:19:61: error: cannot call value of non-function type 'XCTestSuite'
                let darwinCount = thisClass.defaultTestSuite().tests.count
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~^~

/Users/pietbrauer/Projects/RequestKit/Tests/RequestKitTests/JSONPostRouterTests.swift:18:61: error: cannot call value of non-function type 'XCTestSuite'
                let darwinCount = thisClass.defaultTestSuite().tests.count
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~^~

/Users/pietbrauer/Projects/RequestKit/Tests/RequestKitTests/RouterTests.swift:25:61: error: cannot call value of non-function type 'XCTestSuite'
                let darwinCount = thisClass.defaultTestSuite().tests.count
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~^~

pietbrauer committed:
Build on Linux

pietbrauer committed:
Try specifying os

pietbrauer committed:
Make scripts executable

pietbrauer committed:
Make tests compile on Linux

pietbrauer committed:
Fix tests on Linux

tiferrei commented:
BTW this is an ongoing discussion on [Swift Forums](https://forums.swift.org/t/identical-code-compiling-on-macos-but-not-on-linux) as well.

tiferrei committed:
Merge remote-tracking branch 'origin/linux' into linux

tiferrei commented:
@pietbrauer I think it would be a good idea to user Swift 4.1's Docker image for CI as it would reduce build time because we don't have to wait for Swift ENV to download.

pietbrauer commented:
Maybe, the other ones ran through in a minute or so. Don't know why the last one took that long.

tiferrei commented:
Oh, that's alright then, probably just some temporary delays on Travis' networks.

tiferrei commented:
For some reason, the `swift test` isn't passing locally for me. We should probably add a macOS `swift test` to the CI as well.

pietbrauer commented:
I had the same issues but I am fine as long as they run in Xcode.

tiferrei commented:
According to some folks from the Swift Forums, this is also likely a bug in Swift, it should just work no matter the platform.

tiferrei commented:
I think the reason the CI failed on macOS is because it's missing some dependencies, shouldn't the `script/travis-install-macos` install the gems?

pietbrauer committed:
Install bundler

pietbrauer commented:
Done

tiferrei commented:
>Done

@pietbrauer thanks a lot, I’ll see if I can get the tests a bit more stable, the platform specific code looks really hacky.

tiferrei commented:
Yup, it's some kind of bug with Swift 3, my RequestKit based framework, UCLKit, is on Swift 4 and I can use the same code for all the platforms.

pietbrauer commented:
But RequestKit is Swift 4 (#37).

tiferrei commented:
Hmm, this is really weird. The compiler is still specifying it as Swift 3, I caught this in the build logs: `-swift-version 3` and Xcode still suggests me to convert to Swift 4.

tiferrei committed:
Swift 4 ready tests (removes platform mess)

tiferrei committed:
Merge branch 'linux' of https://github.com/nerdishbynature/RequestKit into linux

* 'linux' of https://github.com/nerdishbynature/RequestKit:
  Install bundler

tiferrei committed:
Final tweaks

tiferrei committed:
FIXME: Conditional downcast from 'Error?' to 'NSError' is a bridging conversion

tiferrei commented:
Awesome, it works! Both on Linux and macOS, I think the Test targets were still on Swift 3.

tiferrei commented on Tests/RequestKitTests/TestHelper.swift:
```diff
@@ -26,4 +26,8 @@ internal class Helper {
     }
+
+    public static func getNSError(from error: Error?) -> NSError? {
+        return (error as? NSError)
```
Any clue what's up with `Conditional downcast from 'Error?' to 'NSError' is a bridging conversion; did you mean to use 'as'?`?
I get the warning on Xcode and swift test on macOS but no warnings on Linux.

pietbrauer replied:
It fails on Linux with an compiler error. Don't know if this was because the compiler thought this is still Swift 3. Would love to just use `error as NSError` but seems to be unavailable on Linux.

tiferrei replied:
@pietbrauer The fix for this is currently on the Swift master, until it is released folks from the Swift Forums have suggested using platform checks.

tiferrei replied:
Implemented on 01ae897

tiferrei committed:
Use Swift 4 by default

tiferrei commented on Sources/RequestKit/JSONPostRouter.swift:
```diff
@@ -22,11 +22,11 @@ public extension JSONPostRouter {
-                        userInfo[RequestKitErrorKey] = json as AnyObject?
+                    var userInfo = [String: Any]()
+                    if let data = data, let json = try? JSONSerialization.jsonObject(with: data, options: .mutableContainers) as? [String: Any] {
+                        userInfo[RequestKitErrorKey] = json as Any?
```
I think something's broken here, I am no longer receiving userInfo dicts on my errors.

pietbrauer replied:
On Linux or iOS?

tiferrei replied:
Both, I'm afraid. Here are the [CI logs](https://circleci.com/workflow-run/8b199e66-6841-4e8c-afe1-edc2f6227470).

tiferrei replied:
For some reason the platform checks fixed it 👍

tiferrei committed:
Platform check Error -> NSError casts

tiferrei commented:
All good, ready for review.

pietbrauer commented:
Good Job!
